logstreams/src/test/java/io/zeebe/logstreams/processor/AsyncSnapshotingTest.java

```diff
@@ -178,72 +151,211 @@ public void shouldNotStopTakingSnapshotsAfterFailingReplication() throws Excepti
+  }
+
+  @Test
+  public void shouldNotEnforceSnapshotCreationIfExists() throws Exception {
+    // given
+    final long lastProcessedPosition = 25L;
+    final long lastWrittenPosition = 26L;
+    final long commitPosition = 100L;
+
+    when(lastProcessedEventPositionSupplier.get())
+        .thenReturn(CompletableActorFuture.completed(lastProcessedPosition));
+    when(lastWrittenEventPositionSupplier.get())
+        .thenReturn(CompletableActorFuture.completed(lastWrittenPosition));
+    logStreamRule.setCommitPosition(commitPosition);
+
+    logStreamRule.getClock().addTime(Duration.ofMinutes(1));
+    verify(snapshotController, TIMEOUT).moveValidSnapshot(lastProcessedPosition);
+
+    // when
+    asyncSnapshotDirector
+        .enforceSnapshotCreation(lastWrittenPosition, lastProcessedPosition)
+        .join();
+
+    // then
+    verify(snapshotController, never()).takeSnapshot(lastProcessedPosition);
+  }
+
+  @Test
+  public void shouldNotEnforceSnapshotCreationIfNotCommitted() throws Exception {
+    // given
+    final long lastProcessedPosition = 25L;
+    final long lastWrittenPosition = 101L;
+    final long commitPosition = 100L;
+
+    when(lastProcessedEventPositionSupplier.get())
+        .thenReturn(CompletableActorFuture.completed(lastProcessedPosition));
+    when(lastWrittenEventPositionSupplier.get())
+        .thenReturn(CompletableActorFuture.completed(lastWrittenPosition));
+    logStreamRule.setCommitPosition(commitPosition);
+
+    // when
+    asyncSnapshotDirector
+        .enforceSnapshotCreation(lastWrittenPosition, lastProcessedPosition)
+        .join();
+
+    // then
+    verify(snapshotController, never()).takeSnapshot(lastProcessedPosition);
+  }
+
+  @Test
+  public void shouldNotTakeSnapshotIfExistsAfterRestart() throws IOException {
+    // given
+    final long lastProcessedPosition = 25L;
+    final long lastWrittenPosition = lastProcessedPosition;
+    final long commitPosition = 100L;
+
+    when(lastProcessedEventPositionSupplier.get())
+        .thenReturn(CompletableActorFuture.completed(lastProcessedPosition));
+    when(lastWrittenEventPositionSupplier.get())
+        .thenReturn(CompletableActorFuture.completed(lastWrittenPosition));
+    logStreamRule.setCommitPosition(commitPosition);
+
+    logStreamRule.getClock().addTime(Duration.ofMinutes(1));
+    verify(snapshotController, TIMEOUT).moveValidSnapshot(lastProcessedPosition);
+
+    // when
+    asyncSnapshotDirector.close();
+
+    asyncSnapshotDirector = createAsyncSnapshotDirector();
+    logStreamRule.getActorScheduler().submitActor(asyncSnapshotDirector).join();
+
+    logStreamRule.getClock().addTime(Duration.ofMinutes(1));
+    verify(lastProcessedEventPositionSupplier, TIMEOUT.times(2)).get();
+
+    // then
+    final InOrder inOrder = Mockito.inOrder(snapshotController);
+    inOrder.verify(snapshotController, TIMEOUT.times(1)).getLastValidSnapshotPosition();
```

Comment:
Here I make the same comment as before, we could test more by moving these assertions to before closing the director. The assertions up to line 356 happen before the restart so I would argue that we should verify them before closing the director and then assert that only ```getLastValidSnapshotPosition``` has been invoked after restarting.